libs/shared/form-fields/src/lib/FileUploadController/index.tsx

```diff
@@ -0,0 +1,251 @@
+const answerToUploadFile = (a: UploadFile): UploadFile => {
+  return { name: a.name, key: a.key, url: a.url, status: 'done' }
+}
+
+function reducer(state: UploadFile[], action: Action) {
+  switch (action.type) {
+    case ActionTypes.ADD:
+      return state.concat(action.payload.newFiles)
+
+    case ActionTypes.REMOVE:
+      return state.filter(
+        (file) => file.name !== action.payload.fileToRemove.name,
+      )
+
+    case ActionTypes.UPDATE:
+      return state.map((file: UploadFile) => {
+        if (file.name === action.payload.file.name) {
+          file.status = action.payload.status
+          file.percent = action.payload.percent
+          file.key = action.payload.key
+          file.url = action.payload.url
+        }
+        return file
+      })
+
+    default:
+      return state
+  }
+}
+
+interface FileUploadControllerProps {
+  readonly id: string
+  error?: string
+  application: Application
+  readonly header?: string
+  readonly description?: string
+  readonly buttonLabel?: string
+  readonly multiple?: boolean
+  readonly accept?: string
+  readonly maxSize?: number
+}
+
+export const FileUploadController: FC<FileUploadControllerProps> = ({
+  id,
+  error,
+  application,
+  header,
+  description,
+  buttonLabel,
+  multiple,
+  accept,
+  maxSize,
+}) => {
+  const { clearErrors, setValue } = useFormContext()
+  const [uploadError, setUploadError] = useState<string | undefined>(undefined)
+  const val = getValueViaPath(application.answers, id, []) as UploadFile[]
+
+  const [createUploadUrl] = useMutation(CREATE_UPLOAD_URL)
+
+  const [addAttachment] = useMutation(ADD_ATTACHMENT)
+
+  const [deleteAttachment] = useMutation(DELETE_ATTACHMENT)
+
+  const initialUploadFiles: UploadFile[] =
+    (val && val.map((f) => answerToUploadFile(f))) || []
+
+  const [state, dispatch] = useReducer(reducer, initialUploadFiles)
+
+  useEffect(() => {
+    const onlyUploadedFiles = state.filter(
+      (f: UploadFile) => f.key && f.url && f.status === 'done',
+    )
+
+    const uploadAnswer: UploadFileAnswer[] = onlyUploadedFiles.map(
+      transformToAnswer,
+    )
+
+    setValue(id, uploadAnswer)
+  }, [state, id, setValue])
+
+  const uploadFileFlow = async (file: UploadFile) => {
+    try {
+      // 1. Get the upload URL
+      const { data } = await createUploadUrl({
+        variables: {
+          filename: file.name,
+        },
+      })
+
+      // 2. Upload the file to S3
+      const {
+        createUploadUrl: { url, fields },
+      } = data
+
+      const response = await uploadFileToS3(file, dispatch, url, fields)
+
+      // 3. Add Attachment Data
+      await addAttachment({
+        variables: {
+          input: {
+            id: application.id,
+            key: fields.key,
+            url: `${response.url}/${fields.key}`,
+          },
+        },
+      })
+
+      // Done!
+      return Promise.resolve({ url: response.url, key: fields.key })
+    } catch (e) {
+      error = e
```

Comment:
Wouldn't it be better to have a useState for that with the default value to be the error props and then we can use the setter to pass the new e object as an error?

Reply:
Makes sense thanks, I've refactored it to use the `uploadError` state consistently - with the `error` prop as the default.